qiskit/dagcircuit/dagcircuit.py

```diff
@@ -861,19 +863,89 @@ def idle_wires(self, ignore=None):
                 else:
                     yield wire
 
-    def size(self):
-        """Return the number of operations."""
-        return len(self._multi_graph) - 2 * len(self._wires)
+    def size(self, *, recurse: bool = False):
+        """Return the number of operations.  If there is control flow present, this count may only
+        be an estimate, as the complete control-flow path cannot be staticly known.
+
+        Args:
+            recurse: if ``True``, then recurse into control-flow operations.  For loops with
+                known-length iterators are counted unrolled.  If-else blocks take the longer case of
+                the two branches.  While loops are counted as if the condition runs once only.
+                Defaults to ``False`` and raises :class:`.DAGCircuitError` if any control flow is
+                present, to avoid silently returning a nonsensical number.
+
+        Returns:
+            int: the circuit size
+
+        Raises:
+            DAGCircuitError: if unknown control flow is present in a recursive call, or any control
+                flow is present in a non-recursive call.
+        """
+        length = len(self._multi_graph) - 2 * len(self._wires)
+        if not recurse:
+            if any(x in self._op_names for x in ("for_loop", "while_loop", "if_else")):
+                raise DAGCircuitError(
+                    "Cowardly refusing to give a size for control flow without `recurse=True`."
+                )
+            return length
+        # pylint: disable=cyclic-import
+        from qiskit.converters import circuit_to_dag
+
+        for node in self.op_nodes(ControlFlowOp):
+            if isinstance(node.op, ForLoopOp):
+                indexset = node.op.params[0]
+                inner = len(indexset) * circuit_to_dag(node.op.blocks[0]).size(recurse=True)
+            elif isinstance(node.op, WhileLoopOp):
+                inner = circuit_to_dag(node.op.blocks[0]).size(recurse=True)
+            elif isinstance(node.op, IfElseOp):
+                inner = max(circuit_to_dag(block).size(recurse=True) for block in node.op.blocks)
+            else:
+                raise DAGCircuitError(f"unknown control-flow type: '{node.op.name}'")
+            # Replace the "1" for the node itself with the actual count.
+            length += inner - 1
+        return length
+
+    def depth(self, *, recurse: bool = False):
+        """Return the circuit depth.  If there is control flow present, this count may only be an
+        estimate, as the complete control-flow path cannot be staticly known.
+
+        Args:
+            recurse: if ``True``, then recurse into control-flow operations.  For loops
+                with known-length iterators are counted unrolled.  If-else blocks take the
```

Comment:
```suggestion
                with known-length iterators are unrolled and the resulting depth is counted.  If-else blocks take the
```
Possibly clearer? Trying to avoid ambiguity around whether or not the original `ForLoopOp` will be counted if `recurse=True`, but maybe there's a better wording.

Reply:
Done (something similar) in 312ff332b.